### PR TITLE
Add support for Deepseek V2

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

Deepseek V2 is a MoE model from Deepseek. Relevant variations compared to other models:

- Grouped top-K in expert selection.
- mscale in yarn is calculated using the `mscale` and `mscale_all_dim` configuration options.
- `mscale_all_dim` is also used in scaling attention softmax.
- Permuting of the query/key representations before applying rotary embeddings.
- Some projections cannot be sharded (`q_a_proj`, `kv_a_proj_with_mqa`). So, we need weight loads that supports quantized weights. To this end `{Weights,WeightsLoader}.get_weight` was added.
- The query/key head dimensionality differs from that of the value, so we need to pad during attention.
- Heads with size 192, needs an extension to our paged attention fork and we need to ensure that the KV cache is allocated with the correct size.
- Shared experts.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by drbh — CHANGES_REQUESTED** [review]
looks great just a couple small syntax comments

File: server/Makefile-vllm [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 		git clone https://github.com/Narsil/vllm.git vllm; \
 	fi
-	cd vllm  && git fetch && git checkout $(commit_cuda) && python setup.py build
+	# Remove pull/2/head:pr-2 once https://github.com/Narsil/vllm/pull/2 is merged.
+	cd vllm  && git fetch origin pull/2/head:pr-2 && git checkout $(commit_cuda) && python setup.py build
 
 install-vllm-cuda: build-vllm-cuda
-	cd vllm  && git fetch && git checkout $(commit_cuda) && pip install -e .
+	# Remove pull/2/head:pr-2 once https://github.com/Narsil/vllm/pull/2 is merged.
+	cd vllm  && git fetch origin pull/2/head:pr-2 && git checkout $(commit_cuda) && pip install -e .
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/Makefile-vllm]: Ah cool, I didn't see that. Updated.

File: server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py [hunk 2]
```diff
@@ -0,0 +1,1003 @@
...
+            bos_token_id=bos_token_id,
+            eos_token_id=eos_token_id,
+            tie_word_embeddings=tie_word_embeddings,
+            **kwargs,
+        )
+
+        super().__init__(
+            tie_word_embeddings=tie_word_embeddings,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py]: Fixed.

File: server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py [hunk 3]
```diff
@@ -0,0 +1,1003 @@
...
+
+        self.softmax_scale = self.head_size**-0.5
+
+        from text_generation_server.layers.rotary import get_mscale
+
+        mscale = get_mscale(
+            self.rotary_emb.scaling_factor, self.rotary_emb.mscale_all_dim
+        )
+        self.softmax_scale = self.softmax_scale * mscale * mscale
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py]: Moved the import to the top of the file, put the softmax scale calculation in a single line.

File: server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py [hunk 4]
```diff
@@ -0,0 +1,1003 @@
...
+
+
+class BlockSparseMoE(nn.Module):
+    def __init__(self, prefix, config: DeepseekV2Config, weights):
+        super().__init__()
+
+        if config.ep_size != 1:
+            raise ValueError(
+                f"Currently only ep_size == 1 is supported by BlockSparseMoE, was {config.ep_size}"
+            )
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py]: Done.

File: server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py [hunk 5]
```diff
@@ -0,0 +1,1003 @@
...
+        topk_weights, topk_ids = grouped_topk(
+            x,
+            router_logits,
+            self.top_k,
+            renormalize=self.norm_topk_prob,
+            num_expert_group=self.n_expert_group,
+            topk_group=self.topk_group,
+        )
+
+        # out = self.moe_infer_cpu(x, topk_ids, topk_weights) * self.routed_scaling_factor
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_deepseek_v2_modeling.py]: Removed.

**Review by drbh — APPROVED** [review]
lgtm!
